python_transport/wirepas_gateway/protocol/mqtt_wrapper.py

```diff
@@ -0,0 +1,217 @@
+
+        try:
+            self._client.connect(host, port, keepalive=60)
+        except (socket.gaierror, ValueError) as e:
+            self.logger.error("Cannot connect to mqtt {}".format(e))
+            exit(-1)
+
+        # Set options to initial socket
+        self._client.socket().setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 2048)
+
+        self._publish_queue = SelectableQueue()
+
+    def _on_connect(self, client, userdata, flags, rc):
+        if rc != 0:
+            self.logger.error("MQTT cannot connect {}".format(rc))
+            return
+
+        if self.on_connect_cb is not None:
+            self.on_connect_cb()
+
+    def _do_select(self, sock):
+        # Select with a timeout of 1 sec to call loop misc from time to time
+        r, w, e = select(
+            [sock, self._publish_queue],
+            [sock] if self._client.want_write() else [],
+            [],
+            1,
+        )
+
+        # Check if we have something to publish
+        if self._publish_queue in r:
+            try:
+                # Publish everything. Loop is not necessary as
+                # next select will exit immediately if queue not empty
+                while True:
+                    topic, payload, qos, retain = self._publish_queue.get()
+                    self._client.publish(topic, payload, qos=qos, retain=retain)
+
+                    # FIX: read internal sockpairR as it is written but
+                    # never read as we don't use the internal paho loop
+                    # but we have spurious timeout / broken pipe from
+                    # this socket pair
+                    try:
+                        self._client._sockpairR.recv(1)
+                    except Exception:
+                        # This socket is not used at all, so if something is wrong,
+                        # not a big issue. Just keep going
+                        pass
+
+            except TimeoutError:
+                self.logger.debug("Timeout to send payload: {}".format(payload))
+                # In theory, mqtt client shouldn't loose the last packet
+                # If it is not the case, following line could be uncommented
+                # self._publish_queue.put((topic, payload, qos, retain))
+                raise
+            except queue.Empty:
+                # No more packet to publish
+                pass
+
+        if sock in r:
+            self._client.loop_read()
+
+        if sock in w:
+            self._client.loop_write()
+
+        self._client.loop_misc()
+
+    def _get_socket(self):
+        sock = self._client.socket()
+        if sock is not None:
+            return sock
+
+        self.logger.error("MQTT, unexpected disconnection")
+
+        # Socket is not opened anymore, try to reconnect
+        while True:
+            try:
+                self._client.reconnect()
+                break
+            except Exception:
+                # Retry to connect in 1 sec
+                sleep(1)
```

Comment:
Have you considered adding a counter and exiting after a few attempts?

Reply:
Yes good idea. But exiting here mean exiting the full transport module. It is what we want?

Reply:
yes